ksql-engine/src/main/java/io/confluent/ksql/streams/StreamsUtil.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.confluent.ksql.streams;
+
+import static org.apache.kafka.streams.StreamsConfig.OPTIMIZE;
+import static org.apache.kafka.streams.StreamsConfig.TOPOLOGY_OPTIMIZATION;
+
+import io.confluent.ksql.util.KsqlConfig;
+import java.util.Map;
+import java.util.Objects;
+
+public final class StreamsUtil {
+  private StreamsUtil() {
+  }
+
+  public static boolean useProvidedName(final KsqlConfig ksqlConfig) {
+    final Map<String, Object> streamsConfigs = ksqlConfig.getKsqlStreamConfigProps();
+    return Objects.equals(streamsConfigs.get(TOPOLOGY_OPTIMIZATION), OPTIMIZE);
```

Comment:
At a high level, the reasoning here is that state-store and change-log names are by default chosen by streams. In the absence of streams optimizations and changes to the topology itself, the name chosen is guaranteed to be stable. With optimizations enabled, the default name chosen by streams might change across versions that implement different optimizations. 

Implementation-wise, this is because the name is chosen using an internal counter that isn't incremented when streams can optimize away some things. IIRC, the optimizations themselves allow streams to get rid of some change-log and repartition topics (though I'm struggling a bit to think of examples where streams can optimize away repartitions without breaking the app, since the data sitting in the repartition topics at upgrade time needs to be processed. cc @guozhangwang who explained this stuff to me originally), which will change the naming of all downstream changelog and repartition topics that are actually required.

So to turn this optimization setting on, we need to use explicit naming.

It also gives us some wiggle room to optimize the topologies we generate if we want to later on.

Reply:
That is correct. Explicit naming is suggested so that moving forward when new versions streams may generate different topologies it would not break compatibility.

As of the first upgrade, there is a related section added recently to explain what are user's options: https://github.com/confluentinc/docs/blob/master/streams/upgrade-guide.rst#upgrading-your-kafka-streams-applications-to-cp-510

Reply:
Would it not make sense then to have a `CompatibilityBreakingChange` that switches KSQL to use named statestores for all new queries, regardless of stream optimizations or not?

Reply:
Good question, I think it may be worthy to do given we do not have many users in production and the "piss off" cost may be small to bite in.